Microsoft.Toolkit.Mvvm/ComponentModel/ObservableValidator.cs

```diff
@@ -0,0 +1,159 @@
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.ComponentModel.DataAnnotations;
+using System.Diagnostics.Contracts;
+using System.Linq;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Toolkit.Mvvm.ComponentModel
+{
+    /// <summary>
+    /// A base class for objects implementing the <see cref="INotifyDataErrorInfo"/> interface. This class
+    /// also inherits from <see cref="ObservableObject"/>, so it can be used for observable items too.
+    /// </summary>
+    public abstract class ObservableValidator : ObservableObject, INotifyDataErrorInfo
+    {
+        /// <summary>
+        /// The <see cref="Dictionary{TKey,TValue}"/> instance used to store previous validation results.
+        /// </summary>
+        private readonly Dictionary<string, List<ValidationResult>> errors = new Dictionary<string, List<ValidationResult>>();
+
+        /// <summary>
+        /// Indicates the total number of properties with errors (not total errors).
+        /// This is used to allow <see cref="HasErrors"/> to operate in O(1) time, as it can just
+        /// check whether this value is not 0 instead of having to traverse <see cref="errors"/>.
+        /// </summary>
+        private int totalErrors;
+
+        /// <inheritdoc/>
+        public event EventHandler<DataErrorsChangedEventArgs>? ErrorsChanged;
+
+        /// <inheritdoc/>
+        public bool HasErrors => this.totalErrors > 0;
+
+        /// <inheritdoc/>
+        [Pure]
+        public IEnumerable GetErrors(string? propertyName)
+        {
+            // Entity-level errors when the target property is null or empty
+            if (string.IsNullOrEmpty(propertyName))
+            {
+                return this.GetAllErrors();
+            }
+
+            // Property-level errors, if any
+            if (this.errors.TryGetValue(propertyName!, out List<ValidationResult> errors))
+            {
+                return errors;
+            }
+
+            // The INotifyDataErrorInfo.GetErrors method doesn't specify exactly what to
+            // return when the input property name is invalid, but given that the return
+            // type is marked as a non-nullable reference type, here we're returning an
+            // empty array to respect the contract. This also matches the behavior of
+            // this method whenever errors for a valid properties are retrieved.
+            return Array.Empty<ValidationResult>();
+        }
+
+        /// <summary>
+        /// Implements the logic for entity-level errors gathering for <see cref="GetErrors"/>.
+        /// </summary>
+        /// <returns>An <see cref="IEnumerable"/> instance with all the errors in <see cref="errors"/>.</returns>
+        [Pure]
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private IEnumerable GetAllErrors()
+        {
+            return this.errors.Values.SelectMany(errors => errors);
```

Comment:
What do we do for properties marked **Required**?

If those fields are null, should we consider them errors?

Reply:
So, I have to say I'm not 100% I understand the official documentation here.
From the [MS docs](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.inotifydataerrorinfo.geterrors?view=netcore-3.1#System_ComponentModel_INotifyDataErrorInfo_GetErrors_System_String_):

> `GetErrors`: Gets the validation errors for a specified property or for the entire entity.

I read that as "get/retrieve the _currently stored_ errors for either a specific property or for the entity (so for all properties)", as in, return the errors that have previously been computed, without actually doing any validation now. By that logic, the code right now is just returning the existing errors corresponding to the input query 🤔

I mean, that makes sense to me since we expose a separate method to actually validate properties and store/update errors if needed, which is `ValidateProperty`, but I have to say with how the docs are currently phrased that's not entirely clear.

I guess the question is - should `GetErrors` just return existing errors for a given query, or actually perform a new validation for a given property (or all properties, if the whole entity is targeted)? Thoughts?

Reply:
I agree, the docs aren't clear. I imagine, like many things, it's up to us to define the behavior.

How are my properties validated? Are there any samples to take a look at?

Reply:
Some bug in GitHub is causing me to not be able to comment on this further down. But basically, I don't feel like a `ValidateProperty` method should be required. Can `ObservableValidator` override the `SetProperty` method and call `ValidateProperty` itself?

Reply:
But, at the end of the day - I think this ok. When a user opens up a form, we don't want a bunch of errors for fields they haven't entered anything into yet

Reply:
Can we fold `ValidateProperty` into the `SetProperty` method? Those should always go hand-in-hand.

Reply:
Done in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3429/commits/28b1977bd3f0f88f9e8095adbefe69c9d87b54be as discussed on Discord 👍